Quirks are not always correctly computed, for instance for reloaded documents

youennf opened the pull request:
#### e25e4129a517a4ac302807dda682de8fa7b035da
<pre>
Quirks are not always correctly computed, for instance for reloaded documents
<a href="https://bugs.webkit.org/show_bug.cgi?id=298596">https://bugs.webkit.org/show_bug.cgi?id=298596</a>
<a href="https://rdar.apple.com/160197523">rdar://160197523</a>

Reviewed by Brent Fulgham.

Quirks may be determined early in the Document cycle, for instance when parsing is finished and quirks are checked for styling.
At that time, the Document may not be attached to a Frame, which may lead to the Document having an empty top document URL.
This will prevent correct computation of some quirks, which will remain for the lifetime of the document.

To fix this, we recompute quirks when a document becomes a top document.
Manually tested.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::didBecomeCurrentDocumentInFrame):
* Source/WebCore/page/Quirks.h:

Canonical link: <a href="https://commits.webkit.org/299804@main">https://commits.webkit.org/299804@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c8745995bc44c93e1cc69895da263f993ab6781

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120242 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39936 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30587 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126602 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72315 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122118 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40632 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48513 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/91316 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60610 "An unexpected error occured. Recent messages:Printed configuration") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123194 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32448 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107801 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71868 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31481 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25908 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70226 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101928 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26092 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129496 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47162 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35785 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/99933 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47528 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103986 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99775 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25333 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45229 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/23257 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43795 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47025 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52730 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46493 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49839 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48176 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/61a94929b87ef83df0177336e453c696ff3c01ae. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50493

brentfulgham reviewed (approved):
r=me, with a nit: The radar and bugzilla order are reversed in your Commit message. Can you fix that to canonical order (Bugzilla first, radar second on its own line?)

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9c8745995bc44c93e1cc69895da263f993ab6781. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50493

youennf commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3484,6 +3484,9 @@ void Document::didBecomeCurrentDocumentInFrame()
     }
+
+    if (isTopDocument() && m_quirks)
+        m_quirks->determineRelevantQuirks();
```
Done

youennf committed:
Quirks are not always correctly computed, for instance for reloaded documents
https://bugs.webkit.org/show_bug.cgi?id=298596
rdar://160197523

Reviewed by Brent Fulgham.

Quirks may be determined early in the Document cycle, for instance when parsing is finished and quirks are checked for styling.
At that time, the Document may not be attached to a Frame, which may lead to the Document having an empty top document URL.
This will prevent correct computation of some quirks, which will remain for the lifetime of the document.

To fix this, we recompute quirks when a document becomes a top document.
Manually tested.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::didBecomeCurrentDocumentInFrame):
* Source/WebCore/page/Quirks.h:

Canonical link: https://commits.webkit.org/299804@main
